src/uipath/_cli/_evals/_runtime.py

```diff
@@ -390,3 +462,317 @@ async def cleanup(self) -> None:
+        """Update progress for completed evaluations."""
+        async with self.lock:
+            self.completed_evals += count
+            progress = (self.completed_evals / self.total_evals) * 100
+            await self.event_bus.publish(
+                EvaluationEvents.EVAL_PROGRESS,
+                EvalProgressEvent(
+                    execution_id=self.execution_id,
+                    completed_evals=self.completed_evals,
+                    total_evals=self.total_evals,
+                    progress_percent=progress,
+                ),
+                wait_for_completion=False,
+            )
+
+    async def update_batch_progress(self, batch_id: int):
+        """Update progress for completed batch."""
+        async with self.lock:
+            self.completed_batches += 1
+            await self.event_bus.publish(
+                EvaluationEvents.BATCH_PROGRESS,
+                BatchProgressEvent(
+                    execution_id=self.execution_id,
+                    completed_batches=self.completed_batches,
+                    total_batches=self.total_batches,
+                    batch_id=batch_id,
+                ),
+                wait_for_completion=False,
+            )
+
+
+class ParallelEvalExecutor(Generic[T, C]):
+    """Executor for running evaluations in parallel batches."""
+
+    def __init__(
+        self,
+        max_workers: int,
+        batch_size: int,
+        factory: UiPathRuntimeFactory[T, C],
+        event_bus: EventBus,
+        context: UiPathEvalContext,
+        evaluators: List[BaseEvaluator[Any]],
+        span_exporter: ExecutionSpanExporter,
+    ):
+        self.max_workers = max_workers
+        self.batch_size = batch_size
+        self.factory = factory
+        self.event_bus = event_bus
+        self.context = context
+        self.evaluators = evaluators
+        self.span_exporter = span_exporter
+        self.semaphore = asyncio.Semaphore(max_workers)
+        self.progress_tracker = None
+
+    def _create_batches(self, eval_items: List[EvaluationItem]) -> List[EvalBatch]:
+        """Create batches from evaluation items."""
+        batches = []
+        for i in range(0, len(eval_items), self.batch_size):
+            batch_items = eval_items[i : i + self.batch_size]
+            batches.append(
+                EvalBatch(batch_id=i // self.batch_size, eval_items=batch_items)
+            )
+        return batches
+
+    async def execute_parallel(self, evaluation_set: EvaluationSet) -> UiPathEvalOutput:
+        """Execute evaluations in parallel batches."""
+        batches = self._create_batches(evaluation_set.evaluations)
+
+        self.progress_tracker = ProgressTracker(
+            execution_id=self.context.execution_id or "",
+            total_evals=len(evaluation_set.evaluations),
+            total_batches=len(batches),
+            event_bus=self.event_bus,
+        )
+
+        tasks = []
+        for batch in batches:
```

Comment:
This is pretty inefficient as a single evaluation can significantly slow down the batch processing.

My suggestion is to completely ditch the batch concept. I.e; just submit individual evaluation item to `execute_evaluation_with_semaphore`. In other words, always use batch size=1 and num_workers = batch_size.